fn_grpc_interface/fn_grpc_interface/components/function_grpc.py

```diff
@@ -0,0 +1,195 @@
+
+                grpc_interface_module_list = grpc_helper_obj.get_grpc_interface_module(module_files,
+                                                                                       _grpc_interface_file_dir,
+                                                                                       _grpc_package_name)
+                log.debug(grpc_interface_module_list)
+
+                grpc_stub_tuple = grpc_helper_obj.get_grpc_class(grpc_interface_module_list, 'stub')
+                if not grpc_stub_tuple:
+                    raise ValueError("No gRPC 'stub' class found in the protocol buffer files in the interface_dir")
+
+                grpc_request_tuple = grpc_helper_obj.get_grpc_class(grpc_interface_module_list,
+                                                                    _grpc_request_method_name.lower())
+            except Exception as error_msg:
+                raise ValueError("Failed to load the gRPC protocol buffer files:\nERROR: {0}".format(error_msg))
+
+            # Converting resilient function input data into json object
+            try:
+                grpc_function_json_data = json.loads(grpc_function_data)
+            except Exception:
+                raise FunctionError(u"Function input 'grpc_function_data' must be a JSON formatted String: {0}".format(grpc_function_json_data))
+
+            # Checking Communication type
+            if _grpc_communication_type.lower().strip() == "unary":
+                try:
+                    # Connecting to gRPC Server with given Authentication type
+                    stub_class_obj, grpc_channel_obj = grpc_helper_obj.gRPC_Connect_Server(_grpc_secure_connection,
+                                                                                           grpc_channel,
+                                                                                           grpc_stub_tuple[1],
+                                                                                           _grpc_certificate_path)
+                    stub_method = grpc_helper_obj.get_method_from_stub_object(stub_class_obj, _grpc_communication_type,
+                                                                              stub_name=_grpc_rpc_stub_method_name)
+                    if grpc_function_json_data:
+                        if stub_method[1] is not None:
+                            if six.PY2:
+                                # convert unicode strings to utf-8 for python 2.7
+                                grpc_function_json_data_tmp = ast.literal_eval(json.dumps(grpc_function_json_data))
+                            else:
+                                grpc_function_json_data_tmp = grpc_function_json_data
+
+                            response_received_tmp = stub_method[1](grpc_request_tuple[1](**grpc_function_json_data_tmp))
+
+                            # Closing the gRPC Server Connection
+                            grpc_channel_obj.close()
```

Comment:
This is not safe. This line will not be reached if line 132 to 148 fails and the gRPC channel will remain open.

We should always clean up and close connections in the `finally` clause of a `try/except/finally` block 

See: https://docs.python.org/3/tutorial/errors.html#defining-clean-up-actions

Reply:
ok